### ISSUE TITLE
Huge GPU memory consumption when using DeepSpeed

### ISSUE THREAD

**Issue by ex4sperans** [lucidrains/DALLE-pytorch#161]
So I decided to try the recently introduced DeepSpeed training on 4x V100 GPUs. Previously I was training on a single T4 or V100 (both have 16GB RAM) with certain model parameters (1024 model dim, 16 layers, 8 heads, 64 head dim if that's important). I was able to use a batch size of 2 with this configuration (256px images and no texts). 

I tried to launch distributed training with DeepSpeed with the same model parameters, but to my surprise, it gave an OOM error.  Using binary search I found that it's possible to train the model using DeepSpeed but only when I set the number of layers to 4 (4x reduction) and use only a single sample per batch per GPU (so, 2x reduction). 

Am I missing something? I'm using the latest master branch with some minor changes that are very unlikely to cause such behavior. 

Any help/suggestion is very much appreciated!

**Comment by afiaka87** [comment]
This worked for me - https://github.com/lucidrains/DALLE-pytorch/pull/158

Added a call to `torch.cuda.empty_cache()` before running training and each time the rank zero GPU is run.

**Comment by ex4sperans** [comment]
Oh wow! That worked, thank you very much! 

So it's rather a DeepSpeed bug/issue?

**Comment by afiaka87** [comment]
Yeah no problem - I think so yeah. Library has a lot of quirks unfortunately. 

Something else which may help improve the usage on each GPU is deep speed's ZeRO stage 1 config. 

Add this file to your project directory:

> ds_zero_stage_1.json
```json
{
    "zero_optimization": {
    "stage":1,
    "reduce_bucket_size": 5e8
  }
}
```

Then you can run train_dalle.py via deepspeed - making sure to include your configuration:
For instance (check parameters):
`deepspeed train_dalle.py --taming --image_text_folder=./dataset --deepspeed --deepspeed_config=./ds_zero_stag
e_2.json`

**Comment by ex4sperans** [comment]
Thank you for your suggestions!

**Comment by afiaka87** [comment]
@ex4sperans no problem - keep an eye on this issue. I'm going to leave it open as we've decided empyting the cache is probably not a great long term solution.  @janEbert I think this is a good place to discuss these efforts.

**Comment by janEbert** [comment]
Hey, I really wish I was able to reproduce this to figure out how to solve this in a clean way but I can't.
Could you try to remove the `torch.cuda.clear_cache()` calls and instead call `torch.cuda.synchronize()` right before the loop?

I'm not sure if it helps but [that's what an official DeepSpeed example uses](https://github.com/microsoft/DeepSpeedExamples/blob/590364d482b592c3a8a44c28141a8139c7918c55/gan/gan_deepspeed_train.py#L113) which also calls the unwrapped model a lot. I'd really appreciate if you could try it out and report the result. Otherwise, I'd suggest filing an issue to the DeepSpeed repository so they can offer more help or even fix the issue.

**Comment by janEbert** [comment]
One more thing about the JSON configuration file – please only modify the `deepspeed_config` dictionary in the `train_{dalle,vae}.py` scripts. The JSON file is ignored when the dictionary is given (at least in the current master).

**Comment by afiaka87** [comment]
> One more thing about the JSON configuration file – please only modify the `deepspeed_config` dictionary in the `train_{dalle,vae}.py` scripts. The JSON file is ignored when the dictionary is given (at least in the current master).

Ah yes - this was before you had implemented that I believe. I'll edit my comment above.

**Comment by afiaka87** [comment]
@janEbert @afiaka87

So yeah, this definitely doesn't help runtimes. 

My `SamplesPerSec` as reported by deepspeed (I think)
**With** `torch.cuda.empty_cache()` calls:
```
SamplesPerSec=8.545269005227627
```

**Without** `torch.cuda.empty_cache()` calls:
```
SamplesPerSec=19.42075868400607
```

**Comment by janEbert** [comment]
Ouch... :D
Well, we definitely need to fix **that**!

Were you able to try using `torch.cuda.synchronize()` before the loop after removing the `empty_cache` calls?

**Comment by afiaka87** [comment]
> Ouch... :D
> Well, we definitely need to fix **that**!
> 
> Were you able to try using `torch.cuda.synchronize()` before the loop after removing the `empty_cache` calls?

That's what I have in my code currently. Unfortunately this bug is tough to reproduce - I recall needing a 8xV100 instance to get it to happen. The synchronize certainly doesn't make things worse - as far as I can tell. Anyway, I say for now we just change it to use 

`torch.cuda.synchronize()` just before training (as in the example you posted) 

and remove both calls to

`torch.cuda.empty_cache()`.

Hopefully if people have the issue again they'll come here - and maybe it will fix it.

**Comment by afiaka87** [comment]
@ex4sperans if you're still around - we could actually use your help in reproducing this issue if you've got the same compute available.

**Comment by janEbert** [comment]
So we just experiment with it? I don't dislike that idea. :)
Will you do the PR?

**Comment by afiaka87** [comment]
https://github.com/lucidrains/DALLE-pytorch/pull/174

@janEbert I added an info message directing them to this issue and left the call in the inner loop commented out. Best we can do for now I think.

**Comment by ex4sperans** [comment]
In my case `torch.cuda.empty_cache()` works well, but with just `torch.cuda.synchronize()` I can't even fit a single sample on a GPU so it clearly doesn't solve this issue.

Can do more experiments following your suggestions.

**Comment by afiaka87** [comment]
@ex4sperans Thanks for the confirmation - I can't get synchronize working either and am apparently having this issue on a "typical" 4xV100 instance. For now I've made another pull request which only calls `torch.cuda.empty_cache()` once before training and on the first iteration.

@ex4sperans Does adding the call above training fix the issue for you? Or do you need it inside the training loop as well?

https://github.com/lucidrains/DALLE-pytorch/pull/175

**Comment by janEbert** [comment]
Thanks for testing, both of you! Too bad it doesn't help. I'd also still suggest filing an issue at the DeepSpeed repo; it's probably on them.

**Comment by ex4sperans** [comment]
> For now I've made another pull request which only calls `torch.cuda.empty_cache()` once before training and on the first iteration.

This works for me and also works a bit faster (compared to calling `empty_cache()` on each iteration). Thanks!

**Comment by afiaka87** [comment]
Thanks @ex4sperans! Super hard to test this stuff ha. 

I'll change the PR to just have the empty_cache call before the training loop -

In terms of hacky fixes - compared to installing cudatoolkit=10.1 and pytorch=1.6.0 just to get sparse attention working, this isn't so bad.  I'll leave this issue up but this is less a priority for me now that it's not killing the runtime.

**Comment by ex4sperans** [comment]
It seems that the issue is even more apparent when switching from 4x V100 to 8x V100. Now it goes OOM even before training - I can't even move the model to GPU. The model is the same and it works if I limit the number of GPUs to 4 on the same 8 GPU machine.

Any known workarounds?

**Comment by janEbert** [comment]
Hey, please [post an issue to the DeepSpeed repo](https://github.com/microsoft/DeepSpeed/issues/).

I can try to look into the source myself and PR to DeepSpeed in case I find something, but it's unlikely to yield results since I can't reproduce the bug, so debugging is kind of hard.
You'll need to provide more information though:
- When exactly does the error occur?
- Is it deterministic (does it always happen)?
- What's the stacktrace?

**Comment by ex4sperans** [comment]
Thanks, I'll post an in issue there!

**Comment by ex4sperans** [comment]
Actually, it seems like this can be a fix: https://github.com/microsoft/DeepSpeed/issues/154

I'm going to test it ASAP

**Comment by janEbert** [comment]
What a great find! That's probably it, thank you so much for looking into it!

**Comment by janEbert** [comment]
Would you like to PR the fix? Setting PyTorch's default CUDA device after DeepSpeed's `init_distributed` call should do the trick.

**Comment by afiaka87** [comment]
> Actually, it seems like this can be a fix: [microsoft/DeepSpeed#154](https://github.com/microsoft/DeepSpeed/issues/154)
> 
> I'm going to test it ASAP

Thanks @ex4sperans and @janEbert! I can now easily run a batch size of 200 instead of the 128 i was limited to before this change!

**Comment by ex4sperans** [comment]
So I was able to run it on 8 V100s just by removing the line `dalle = dalle.cuda()` and relying fully on DeepSpeed to move the model to the correct device.

**Comment by janEbert** [comment]
That sounds nice, too. [A DeepSpeed example](https://github.com/microsoft/DeepSpeedExamples/blob/master/gan/gan_deepspeed_train.py) seems to vouch that the solution I suggested works as well.
I personally like hiding away the work behind the backend initialization function; this also catches possible future errors instead of treating one case specially.

So are you interested in PRing?

**Comment by janEbert** [comment]
I now think your solution of letting DeepSpeed handle CUDA is the better one. They also do not do this in the MegatronLM ZeRO 3 example (where they also don't move the model to half-precision manually) which is probably more relevant for us than the GAN example.
